packages/next/src/client/components/router-reducer/reducers/server-reference-info.ts

```diff
@@ -0,0 +1,75 @@
+export interface ServerReferenceInfo {
+  type: 'server-action' | 'use-cache'
+  usedArgs: [boolean, boolean, boolean, boolean, boolean, boolean]
+  hasRestArgs: boolean
+}
+
+/**
+ * Extracts info about the server reference for the given server reference ID by
+ * parsing the first byte of the hex-encoded ID.
+ *
+ * ```
+ * Bit positions: [7]      [6] [5] [4] [3] [2] [1]  [0]
+ * Bits:          typeBit  argMask                  restArgs
+ * ```
+ *
+ * If the `typeBit` is `1` the server reference represents a `"use cache"`
+ * function, otherwise a server action.
+ *
+ * The `argMask` encodes whether the function uses the argument at the
+ * respective position.
+ *
+ * The `restArgs` bit indicates whether the function uses a rest parameter.
+ *
+ * @param id hex-encoded server reference ID
+ */
+export function extractInfoFromServerReferenceId(
+  id: string
+): ServerReferenceInfo {
+  const infoByte = parseInt(id.slice(0, 2), 16)
+  const typeBit = (infoByte >> 7) & 0x1
+  const argMask = (infoByte >> 1) & 0x3f
+  const restArgs = infoByte & 0x1
+  const usedArgs = Array(6)
+
+  for (let index = 0; index < 6; index++) {
+    const bitPosition = 5 - index
+    const bit = (argMask >> bitPosition) & 0x1
+    usedArgs[index] = bit === 1
+  }
+
+  return {
+    type: typeBit === 1 ? 'use-cache' : 'server-action',
+    usedArgs: usedArgs as [
+      boolean,
+      boolean,
+      boolean,
+      boolean,
+      boolean,
+      boolean,
+    ],
+    hasRestArgs: restArgs === 1,
+  }
+}
+
+/**
+ * Creates a sparse array containing only the used arguments based on the
+ * provided action info.
+ */
+export function filterArgs(
+  args: unknown[],
+  info: ServerReferenceInfo
+): unknown[] {
+  const filteredArgs = new Array(args.length)
+
+  for (let index = 0; index < args.length; index++) {
+    if (
+      (index < 6 && info.usedArgs[index]) ||
+      (index >= 6 && info.hasRestArgs)
```

Comment:
isn't this gonna silently skip stuff if we have a function with 7+ args but no rest arg?

Reply:
Yes, good catch, that needs to be changed as long as we don't fail for these at build time.

Reply:
Actually, we don't need to do that, the transform sets the restArgs bit when there are more than 6 args:
https://github.com/vercel/next.js/blob/c4694048245efbbe0ddec0f3f059b73fe461433b/crates/next-custom-transforms/src/transforms/server_actions.rs#L196-L198

Reply:
hmmmm. honestly, i don't like that very much, feels like `restArgs` should mean something, but sure, i guess it works

i'd consider adding an invariant here like
```
if (args.length > 6 && !info.hasRestArgs) {
  throw new InvariantError('oops')
}
```

or at least commenting that we're relying on this behavior

Reply:
I don't think we can throw the invariant error, because a function might use 6 params but it's called with 7. In this case we do want to omit the 7th. But I did add comments regarding the transform assumption.